### ISSUE TITLE
Panic: "invalid key" in Streams::recv_reset

### ISSUE THREAD

**Issue by jeromegn** [hyperium/h2#362]
We caught this in production via Sentry.io:

```
OS Version: Linux 4.13.0-36-generic (None)
Report Version: 104

Application Specific Information:
invalid key

Thread 0
0   slab                            0x924671            [inlined] <slab::Slab<T> as core::ops::index::Index<T>>::index (<::std::macros::panic macros>:3)
1   h2                              0x924671            <h2::proto::streams::store::Ptr<T> as core::ops::deref::Deref>::deref (store.rs:361)
2   h2                              0x553e4f            [inlined] <h2::proto::streams::streams::Streams<T>>::recv_reset::{{closure}} (streams.rs:294)
3   h2                              0x553e4f            h2::proto::streams::counts::Counts::transition (counts.rs:127)
4   h2                              0x46a47d            <h2::proto::streams::streams::Streams<T>>::recv_reset (streams.rs:291)
5   h2                              0x57397b            [inlined] <h2::proto::connection::Connection<T>>::poll2 (connection.rs:321)
6   h2                              0x57397b            <h2::proto::connection::Connection<T>>::poll (connection.rs:203)
7   h2                              0x54700f            [inlined] <h2::server::Connection<T>>::poll_close (server.rs:424)
8   h2                              0x54700f            <h2::server::Connection<T> as futures::stream::Stream>::poll (server.rs:487)
9   hyper                           0x551884            [inlined] <hyper::proto::h2::server::Serving<T>>::poll_server (server.rs:162)
10  hyper                           0x551884            <hyper::proto::h2::server::Server<T> as futures::future::Future>::poll (server.rs:104)
11  futures                         0x475912            [inlined] <futures::future::either::Either<T> as futures::future::Future>::poll (either.rs:36)
12  futures                         0x475912            [inlined] futures::future::option::<T>::poll (option.rs:12)
13  hyper                           0x475912            <hyper::server::conn::upgrades::UpgradeableConnection<T> as futures::future::Future>::poll (conn.rs:946)
14  hyper                           0x4b9234            <hyper::common::drain::Watching<T> as futures::future::Future>::poll (drain.rs:105)
15  hyper                           0x4777a2            <hyper::server::conn::spawn_all::NewSvcTask<T> as futures::future::Future>::poll (conn.rs:886)
16  alloc                           0xb3dd31            [inlined] <alloc::boxed::Box<T> as futures::future::Future>::poll (mod.rs:113)
17  futures                         0xb3dd31            [inlined] <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}} (mod.rs:326)
18  futures                         0xb3dd31            [inlined] <futures::task_impl::Spawn<T>>::enter::{{closure}} (mod.rs:396)
19  futures                         0xb3dd31            futures::task_impl::std::set (mod.rs:78)
20  futures                         0xb3e825            [inlined] <futures::task_impl::Spawn<T>>::enter (mod.rs:396)
21  futures                         0xb3e825            [inlined] <futures::task_impl::Spawn<T>>::poll_fn_notify (mod.rs:288)
22  futures                         0xb3e825            <futures::task_impl::Spawn<T>>::poll_future_notify (mod.rs:326)
23  tokio_threadpool                0xb41bb0            [inlined] tokio_threadpool::task::Task::run::{{closure}} (mod.rs:145)
24  core                            0xb41bb0            [inlined] core::ops::function::FnOnce::call_once (function.rs:231)
25  std                             0xb41bb0            [inlined] <std::panic::AssertUnwindSafe<T> as core::ops::function::FnOnce<T>>::call_once (panic.rs:309)
26  std                             0xb41bb0            std::panicking::try::do_call (panicking.rs:297)
27  <unknown>                       0xcb22c9            __rust_maybe_catch_panic (lib.rs:87)
28  std                             0xb41eff            [inlined] std::panicking::try (panicking.rs:276)
29  std                             0xb41eff            [inlined] std::panic::catch_unwind (panic.rs:388)
30  tokio_threadpool                0xb41eff            tokio_threadpool::task::Task::run (mod.rs:130)
31  tokio_threadpool                0xb3a644            [inlined] tokio_threadpool::worker::Worker::run_task2 (mod.rs:567)
32  tokio_threadpool                0xb3a644            tokio_threadpool::worker::Worker::run_task (mod.rs:459)
33  tokio_threadpool                0xb39b3d            [inlined] tokio_threadpool::worker::Worker::try_run_owned_task (mod.rs:390)
34  tokio_threadpool                0xb39b3d            [inlined] tokio_threadpool::worker::Worker::try_run_task (mod.rs:297)
35  tokio_threadpool                0xb39b3d            tokio_threadpool::worker::Worker::run (mod.rs:241)
36  tokio                           0xb13d26            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}}::{{closure}} (builder.rs:349)
37  tokio_trace_core                0xb13d26            tokio_trace_core::dispatcher::with_default (dispatcher.rs:54)
38  tokio                           0xb152e5            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}} (builder.rs:348)
39  tokio_timer                     0xb152e5            [inlined] tokio_timer::timer::handle::with_default::{{closure}} (handle.rs:94)
40  std                             0xb152e5            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
41  std                             0xb152e5            <std::thread::local::LocalKey<T>>::with (local.rs:246)
42  tokio_timer                     0xb154ad            [inlined] tokio_timer::timer::handle::with_default (handle.rs:81)
43  tokio                           0xb154ad            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}} (builder.rs:347)
44  tokio_timer                     0xb154ad            [inlined] tokio_timer::clock::clock::with_default::{{closure}} (clock.rs:141)
45  std                             0xb154ad            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
46  std                             0xb154ad            <std::thread::local::LocalKey<T>>::with (local.rs:246)
47  tokio_timer                     0xb15753            [inlined] tokio_timer::clock::clock::with_default (clock.rs:124)
48  tokio                           0xb15753            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}} (builder.rs:346)
49  tokio_reactor                   0xb15753            [inlined] tokio_reactor::with_default::{{closure}} (lib.rs:237)
50  std                             0xb15753            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
51  std                             0xb15753            <std::thread::local::LocalKey<T>>::with (local.rs:246)
52  tokio_reactor                   0xb1b2f6            [inlined] tokio_reactor::with_default (lib.rs:217)
53  tokio                           0xb1b2f6            tokio::runtime::threadpool::builder::Builder::build::{{closure}} (builder.rs:345)
54  tokio_threadpool                0xb42ac2            [inlined] tokio_threadpool::callback::Callback::call (callback.rs:22)
55  tokio_threadpool                0xb42ac2            [inlined] tokio_threadpool::worker::Worker::do_run::{{closure}}::{{closure}} (mod.rs:127)
56  tokio_executor                  0xb42ac2            [inlined] tokio_executor::global::with_default::{{closure}} (global.rs:209)
57  std                             0xb42ac2            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
58  std                             0xb42ac2            <std::thread::local::LocalKey<T>>::with (local.rs:246)
59  tokio_executor                  0xb4289d            [inlined] tokio_executor::global::with_default (global.rs:178)
60  tokio_threadpool                0xb4289d            [inlined] tokio_threadpool::worker::Worker::do_run::{{closure}} (mod.rs:125)
61  std                             0xb4289d            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
62  std                             0xb4289d            <std::thread::local::LocalKey<T>>::with (local.rs:246)
63  tokio_threadpool                0xb445f0            [inlined] tokio_threadpool::worker::Worker::do_run (mod.rs:116)
64  tokio_threadpool                0xb445f0            [inlined] tokio_threadpool::pool::Pool::spawn_thread::{{closure}} (mod.rs:344)
65  std                             0xb445f0            std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:135)
66  std                             0xb41bfb            [inlined] std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} (mod.rs:469)
67  std                             0xb41bfb            [inlined] <std::panic::AssertUnwindSafe<T> as core::ops::function::FnOnce<T>>::call_once (panic.rs:309)
68  std                             0xb41bfb            std::panicking::try::do_call (panicking.rs:297)
69  <unknown>                       0xcb22c9            __rust_maybe_catch_panic (lib.rs:87)
70  std                             0xb3c5ff            [inlined] std::panicking::try (panicking.rs:276)
71  std                             0xb3c5ff            [inlined] std::panic::catch_unwind (panic.rs:388)
72  std                             0xb3c5ff            [inlined] std::thread::Builder::spawn_unchecked::{{closure}} (mod.rs:468)
73  alloc                           0xb3c5ff            <F as alloc::boxed::FnBox<T>>::call_box (boxed.rs:749)
74  <unknown>                       0xcb192d            [inlined] call_once<T> (boxed.rs:759)
75  <unknown>                       0xcb192d            [inlined] start_thread (thread.rs:14)
76  <unknown>                       0xcb192d            thread_start (thread.rs:81)
```

This is using h2 0.1.21 (latest as of this writing).

It does not happen often. We serve a few thousand requests per second and this only happened about 6 times in 2 days.

**Comment by seanmonstar** [comment]
Yikes! That it's in `recv_reset`, it sounds like a RET_STREAM frame is received for a stream that we either a) never had before, or b) already finished and removed from our store.

There *should* be checks in that function to protect against that, but seems there's some way to slip through.

**Comment by seanmonstar** [comment]
I *believe* #364 fixes this, but as described there, I couldn't come up with a test exercising it.

**Comment by jeromegn** [comment]
I haven't tried the new commits, but I got a different trace for the same panic just now:

```
OS Version: Linux 4.13.0-41-generic (None)
Report Version: 104

Application Specific Information:
invalid key

Thread 0
0   h2                              0x91ef97            h2::proto::streams::counts::Counts::transition_after
1   h2                              0x5543bd            h2::proto::streams::counts::Counts::transition (counts.rs:129)
2   h2                              0x477a5f            [inlined] <h2::proto::streams::streams::Streams<T>>::recv_go_away::{{closure}} (streams.rs:353)
3   h2                              0x477a5f            h2::proto::streams::store::Store::for_each (store.rs:145)
4   h2                              0x46aef9            <h2::proto::streams::streams::Streams<T>>::recv_go_away (streams.rs:351)
5   h2                              0x573b17            [inlined] <h2::proto::connection::Connection<T>>::poll2 (connection.rs:337)
6   h2                              0x573b17            <h2::proto::connection::Connection<T>>::poll (connection.rs:203)
7   h2                              0x54700f            [inlined] <h2::server::Connection<T>>::poll_close (server.rs:424)
8   h2                              0x54700f            <h2::server::Connection<T> as futures::stream::Stream>::poll (server.rs:487)
9   hyper                           0x551884            [inlined] <hyper::proto::h2::server::Serving<T>>::poll_server (server.rs:162)
10  hyper                           0x551884            <hyper::proto::h2::server::Server<T> as futures::future::Future>::poll (server.rs:104)
11  futures                         0x475912            [inlined] <futures::future::either::Either<T> as futures::future::Future>::poll (either.rs:36)
12  futures                         0x475912            [inlined] futures::future::option::<T>::poll (option.rs:12)
13  hyper                           0x475912            <hyper::server::conn::upgrades::UpgradeableConnection<T> as futures::future::Future>::poll (conn.rs:946)
14  hyper                           0x4b9234            <hyper::common::drain::Watching<T> as futures::future::Future>::poll (drain.rs:105)
15  hyper                           0x4777a2            <hyper::server::conn::spawn_all::NewSvcTask<T> as futures::future::Future>::poll (conn.rs:886)
16  alloc                           0xb3dd31            [inlined] <alloc::boxed::Box<T> as futures::future::Future>::poll (mod.rs:113)
17  futures                         0xb3dd31            [inlined] <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}} (mod.rs:326)
18  futures                         0xb3dd31            [inlined] <futures::task_impl::Spawn<T>>::enter::{{closure}} (mod.rs:396)
19  futures                         0xb3dd31            futures::task_impl::std::set (mod.rs:78)
20  futures                         0xb3e825            [inlined] <futures::task_impl::Spawn<T>>::enter (mod.rs:396)
21  futures                         0xb3e825            [inlined] <futures::task_impl::Spawn<T>>::poll_fn_notify (mod.rs:288)
22  futures                         0xb3e825            <futures::task_impl::Spawn<T>>::poll_future_notify (mod.rs:326)
23  tokio_threadpool                0xb41bb0            [inlined] tokio_threadpool::task::Task::run::{{closure}} (mod.rs:145)
24  core                            0xb41bb0            [inlined] core::ops::function::FnOnce::call_once (function.rs:231)
25  std                             0xb41bb0            [inlined] <std::panic::AssertUnwindSafe<T> as core::ops::function::FnOnce<T>>::call_once (panic.rs:309)
26  std                             0xb41bb0            std::panicking::try::do_call (panicking.rs:297)
27  <unknown>                       0xcb22c9            __rust_maybe_catch_panic (lib.rs:87)
28  std                             0xb41eff            [inlined] std::panicking::try (panicking.rs:276)
29  std                             0xb41eff            [inlined] std::panic::catch_unwind (panic.rs:388)
30  tokio_threadpool                0xb41eff            tokio_threadpool::task::Task::run (mod.rs:130)
31  tokio_threadpool                0xb3a644            [inlined] tokio_threadpool::worker::Worker::run_task2 (mod.rs:567)
32  tokio_threadpool                0xb3a644            tokio_threadpool::worker::Worker::run_task (mod.rs:459)
33  tokio_threadpool                0xb39d48            [inlined] tokio_threadpool::worker::Worker::try_steal_task (mod.rs:416)
34  tokio_threadpool                0xb39d48            [inlined] tokio_threadpool::worker::Worker::try_run_task (mod.rs:301)
35  tokio_threadpool                0xb39d48            tokio_threadpool::worker::Worker::run (mod.rs:241)
36  tokio                           0xb13d26            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}}::{{closure}} (builder.rs:349)
37  tokio_trace_core                0xb13d26            tokio_trace_core::dispatcher::with_default (dispatcher.rs:54)
38  tokio                           0xb152e5            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}} (builder.rs:348)
39  tokio_timer                     0xb152e5            [inlined] tokio_timer::timer::handle::with_default::{{closure}} (handle.rs:94)
40  std                             0xb152e5            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
41  std                             0xb152e5            <std::thread::local::LocalKey<T>>::with (local.rs:246)
42  tokio_timer                     0xb154ad            [inlined] tokio_timer::timer::handle::with_default (handle.rs:81)
43  tokio                           0xb154ad            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}}::{{closure}} (builder.rs:347)
44  tokio_timer                     0xb154ad            [inlined] tokio_timer::clock::clock::with_default::{{closure}} (clock.rs:141)
45  std                             0xb154ad            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
46  std                             0xb154ad            <std::thread::local::LocalKey<T>>::with (local.rs:246)
47  tokio_timer                     0xb15753            [inlined] tokio_timer::clock::clock::with_default (clock.rs:124)
48  tokio                           0xb15753            [inlined] tokio::runtime::threadpool::builder::Builder::build::{{closure}}::{{closure}} (builder.rs:346)
49  tokio_reactor                   0xb15753            [inlined] tokio_reactor::with_default::{{closure}} (lib.rs:237)
50  std                             0xb15753            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
51  std                             0xb15753            <std::thread::local::LocalKey<T>>::with (local.rs:246)
52  tokio_reactor                   0xb1b2f6            [inlined] tokio_reactor::with_default (lib.rs:217)
53  tokio                           0xb1b2f6            tokio::runtime::threadpool::builder::Builder::build::{{closure}} (builder.rs:345)
54  tokio_threadpool                0xb42ac2            [inlined] tokio_threadpool::callback::Callback::call (callback.rs:22)
55  tokio_threadpool                0xb42ac2            [inlined] tokio_threadpool::worker::Worker::do_run::{{closure}}::{{closure}} (mod.rs:127)
56  tokio_executor                  0xb42ac2            [inlined] tokio_executor::global::with_default::{{closure}} (global.rs:209)
57  std                             0xb42ac2            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
58  std                             0xb42ac2            <std::thread::local::LocalKey<T>>::with (local.rs:246)
59  tokio_executor                  0xb4289d            [inlined] tokio_executor::global::with_default (global.rs:178)
60  tokio_threadpool                0xb4289d            [inlined] tokio_threadpool::worker::Worker::do_run::{{closure}} (mod.rs:125)
61  std                             0xb4289d            [inlined] <std::thread::local::LocalKey<T>>::try_with (local.rs:300)
62  std                             0xb4289d            <std::thread::local::LocalKey<T>>::with (local.rs:246)
63  tokio_threadpool                0xb445f0            [inlined] tokio_threadpool::worker::Worker::do_run (mod.rs:116)
64  tokio_threadpool                0xb445f0            [inlined] tokio_threadpool::pool::Pool::spawn_thread::{{closure}} (mod.rs:344)
65  std                             0xb445f0            std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:135)
66  std                             0xb41bfb            [inlined] std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} (mod.rs:469)
67  std                             0xb41bfb            [inlined] <std::panic::AssertUnwindSafe<T> as core::ops::function::FnOnce<T>>::call_once (panic.rs:309)
68  std                             0xb41bfb            std::panicking::try::do_call (panicking.rs:297)
69  <unknown>                       0xcb22c9            __rust_maybe_catch_panic (lib.rs:87)
70  std                             0xb3c5ff            [inlined] std::panicking::try (panicking.rs:276)
71  std                             0xb3c5ff            [inlined] std::panic::catch_unwind (panic.rs:388)
72  std                             0xb3c5ff            [inlined] std::thread::Builder::spawn_unchecked::{{closure}} (mod.rs:468)
73  alloc                           0xb3c5ff            <F as alloc::boxed::FnBox<T>>::call_box (boxed.rs:749)
74  <unknown>                       0xcb192d            [inlined] call_once<T> (boxed.rs:759)
75  <unknown>                       0xcb192d            [inlined] start_thread (thread.rs:14)
76  <unknown>                       0xcb192d            thread_start (thread.rs:81)
```

I'll try the fixed ref soon! Thanks.

**Comment by seanmonstar** [comment]
I took a look at what happens in that new function as well, and I *believe* the fix for the previous should fix this one too. I've published the fix in a new version, 0.1.22, let me know if you see panics after upgrading.
